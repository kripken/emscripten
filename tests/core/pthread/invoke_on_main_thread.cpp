#include <iostream>

#include <emscripten.h>

#include <emscripten/thread_utils.h>

int main() {
  struct Foo {
    void operator()() {
      // Print whether we are on the main thread.
      auto mainThread = EM_ASM_INT({
        var mainThread = typeof importScripts === 'undefined';
        console.log("hello. mainThread=", mainThread);
        return mainThread;
      });

      // If we are on the main thread, it is time to end this test. Do so
      // asynchronously, as if we exit right now then the object Foo() we are
      // called on will not yet be destroyed, which causes a false positive in
      // LSan leak detection.
      if (mainThread) {
        emscripten_async_call(
          [](void*) {
            exit(0);
          },
          nullptr,
          0
        );
      }
    }
  };

  // Call it on this thread.
  Foo()();

  // Call it on the main thread.
  emscripten::invoke_on_main_thread_async(Foo());

  emscripten_exit_with_live_runtime();
}
