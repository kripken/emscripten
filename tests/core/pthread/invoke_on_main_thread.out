hello. mainThread= false
hello. mainThread= true
